src/main/java/vazkii/patchouli/client/book/text/SpanState.java

```diff
@@ -21,6 +22,9 @@
 
 	private final Style baseStyle;
 	private final Deque<Style> styleStack = new ArrayDeque<>();
+        // This keeps a list of all modifications made to the style at each layer,
+        // so that when we replace the base style, we can reconstruct the style stack.
+	private final Deque<SpanPartialState> modifiersStack = new ArrayDeque<>();
```

Comment:
consider pulling this + styleStack into a separate structure that ensures that they're consistent with each other. This seems fine right now but bugs might arise later if someone edits code touching styleStack without also knowing they have to do the same for modifiersStack.